### ISSUE TITLE
Ungrab keyboard when prompt fails

### ISSUE THREAD

**Issue by Fuco1** [xmonad/xmonad-contrib#180]
### Problem Description

I use the `XMonad.Prompt` for various things like playing music (selecting the albums etc).  My code probably isn't 100% correct because sometimes it fails (I don't know where or how) and the keyboard is never ungrabbed (the prompt never "exits" properly).  This means my system becomes unusable and the only thing I can do is restart the computer

I think that the prompt should be more resilient and release the keyboard on error.

I have made the following change to `Prompt.hs`

```diff
diff --git a/XMonad/Prompt.hs b/XMonad/Prompt.hs
index 69eb6bf..c20e722 100644
--- a/XMonad/Prompt.hs
+++ b/XMonad/Prompt.hs
@@ -368,7 +368,11 @@ mkXPromptWithReturn t conf compl action = do
   let hs = fromMaybe [] $ M.lookup (showXPrompt t) hist
       om = (XPSingleMode compl (XPT t)) --operation mode
       st = initState d rw w s om gc fs hs conf numlock
-  st' <- io $ execStateT runXP st
+  st' <- io $ catch
+               (execStateT runXP st)
+               (\(SomeException _) -> do
+                  ungrabKeyboard d currentTime
+                  return $ st { successful = False })
 
   releaseXMF fs
   io $ freeGC d gc
```

but I'm not sure if this is the greatest way to handle things---I'm quite unsure about Haskell error handling still.

Does this make sene to you?  Ideally if someone knows how to do this better we should change the patch or just apply the above if it is fine.

Thanks!

**Comment by pjones** [comment]
@Fuco1 Thanks for the patch!  When I get some more time I'll review it and make sure this bug is fixed properly.

**Comment by f1u77y** [comment]
### A little offtopic

> This means my system becomes unusable and the only thing I can do is restart the computer

You could use [magic keys](https://en.wikipedia.org/wiki/Magic_SysRq_key) for accessing the keyboard if it's fully grabbed by X.

**Comment by geekosaur** [comment]
Only if they were enabled beforehand.

Also... if the keyboard is grabbed by xmonad, then (a) xmonad keys should work as long as it reached the main event loop (if it didn't, just ungrabbing won't help) and (b) mod-q should drop the grab, because the server socket gets closed (unless there is an x.org bug that should be reported upstream).

**Comment by Fuco1** [comment]
Well, I call DBUS in my prompt to query for all available media players (this is wrapped in another library) so it might be locking somewhere there.  As I understand it xmonad is single-threaded so a stall there would also stop the event loop from getting events.

It is quite hard to simulate and it happens rather infrequently (but sometimes three times a day, sometimes not for a month).  I might try to review the DBUS library instead.

**Comment by liskin** [comment]
Fixed by #287, closing.
